### DISCUSSION TITLE
Creating a bridge for class with static methods.

### DISCUSSION THREAD

**Discussion by Chappie74** [ethanblake4/dart_eval#46] [Q&A] [answered]
So I have a [class ](https://github.com/Chappie74/flutter_auto_gui/blob/main/flutter_auto_gui/lib/flutter_auto_gui.dart) i'm looking to bridge. It contains all static methods. 

What i'd like to know is how can I define these methods in the bridge boiler plate code?
I see in the example, there is `methods` in the $declaration. But also the comment in $new specifies that's where all static methods go. 

![image](https://user-images.githubusercontent.com/21108487/190538304-12b334d5-9bdf-44b2-90ba-010cf2a148c7.png)

Can someone please give me some assistance on how to properly do this. Thanks in advance.

**Comment by Chappie74** [comment]
I've been going through it piece by piece and this is what I have so far. 


```dart
import 'package:dart_eval/dart_eval_bridge.dart';
import 'package:dart_eval/src/eval/runtime/runtime.dart';
import 'package:dart_eval/stdlib/core.dart';
import 'package:flutter_auto_gui/flutter_auto_gui.dart';

class $FlutterAutoGUI implements FlutterAutoGUI, $Instance {
  /// Define the compile-time type descriptor as an unresolved type
  static const $type = BridgeTypeRef.spec(
      BridgeTypeSpec('package:example/bridge.dart', 'FlutterAutoGUI'));

  @override
  final FlutterAutoGUI $value;

  $FlutterAutoGUI.wrap(this.$value) : _superclass = $Object($value);
  final $Instance _superclass;

  static const $declaration = BridgeClassDef(
    BridgeClassType($type),
    constructors: {},
    methods: {},
    getters: {},
    setters: {},
    fields: {},
    wrap: true,
  );

  static $Value? $new(Runtime runtime, $Value? target, List<$Value?> args) {
    return $FlutterAutoGUI.wrap(FlutterAutoGUI());
  }

  @override
  $Value? $getProperty(Runtime runtime, String identifier) {
    throw UnimplementedError();
  }

  @override
  // TODO: implement $reified
  get $reified => throw UnimplementedError();

  @override
  // TODO: implement $runtimeType
  int get $runtimeType => throw UnimplementedError();

  @override
  void $setProperty(Runtime runtime, String identifier, $Value value) {
    // TODO: implement $setProperty
  }
}

```

**Reply by Chappie74** [reply]
So i've managed to figure this part out. I think i'm on the right track. 
For BridgeTypeRef how can I declare custom or other non built in types?
Also, the functions in the class are Futures, do I need to do something special with those when declaring ?
![image](https://user-images.githubusercontent.com/21108487/190548773-a9c6cf4d-f2d1-470e-b773-0264b88ad792.png)

**Reply by Chappie74** [reply]
Think i've figured it out.
![image](https://user-images.githubusercontent.com/21108487/190550458-8c788304-29d0-4938-b4ca-8ea1ca9a3e69.png)

**Comment by ethanblake4** [comment]
Hey, nice work! You're definitely on the right track :)

So, a few things.
First off, if you're returning a Future you'll need to use `BridgeTypeSpec('dart:core', 'Future')`. Right now, specifying the inner type of the Future doesn't actually do anything, but you can if you want for future releases.

Next, dart_eval doesn't have standard library bindings for Point, Curve, or Size yet. Luckily that's really easy to do so I'll add those in the next few days and publish a new point release.

You'll also probably want to replace 'package:example/bridge.dart` with your actual package and file name so the import makes sense. If you want to be technical about it and specify an internal URI with /src/, you'll also have to do compiler.addSource and put your top-level library file with the export '...' directives.

After that, all that's left is creating the runtime bindings which for you should be entirely with registerBridgeFunc(). The notable thing there is that you'll have to provide a type mapper when wrapping your returned Futures in $Future, so that it can convert the internal type (e.g. Point to $Point). That's just a parameter on the $Future constructor, where you can specify a closure that does `(val) => $Point.wrap(val)` or similar.

One final thing - you'll want to remove that `$new` runtime binding if you're not putting anything in `constructors`. (You don't need them if you're only using static methods... and in fact, you don't need to implement $Instance or FlutterAutoGUI either or override all those methods. Though it won't hurt to leave those in unlike $new)

**Reply by Chappie74** [reply]
>You'll also probably want to replace 'package:example/bridge.dart` with your actual package and file name so the import makes sense. If you want to be technical about it and specify an internal URI with /src/, you'll also have to do compiler.addSource and put your top-level library file with the export '...' directives.

Would I have to do this if I have the following:
![image](https://user-images.githubusercontent.com/21108487/190790648-52323a51-57d1-43a1-8c3e-152bcbe9192b.png)


>After that, all that's left is creating the runtime bindings which for you should be entirely with registerBridgeFunc(). The notable thing there is that you'll have to provide a type mapper when wrapping your returned Futures in $Future, so that it can convert the internal type (e.g. Point to $Point). That's just a parameter on the $Future constructor, where you can specify a closure that does (val) => $Point.wrap(val) or similar.

Would that be something like this?
![image](https://user-images.githubusercontent.com/21108487/190807375-667642fc-4bf5-4ad5-aeeb-5df4893b38c0.png)
If so, i'm unable to do $value.position

Do I need to define these overloads for the static functions? If so how would I go about doing that?

![image](https://user-images.githubusercontent.com/21108487/190808181-4db0c788-3c1b-4584-a3a9-cfc374a66e87.png)


Finally this is what my file now looks like

```dart
import 'package:dart_eval/dart_eval.dart';
import 'package:dart_eval/dart_eval_bridge.dart';
import 'package:dart_eval/stdlib/core.dart';
import 'package:flutter_auto_gui/flutter_auto_gui.dart';

class $FlutterAutoGUI extends $Instance {
  static const $type = BridgeTypeRef.spec(
    BridgeTypeSpec(
      'package:flutter_auto_gui/flutter_auto_gui.dart',
      'FlutterAutoGUI',
    ),
  );

  @override
  final FlutterAutoGUI $value;

  $FlutterAutoGUI.wrap(this.$value) : _superclass = $Object($value);
  final $Instance _superclass;

  static const $declaration = BridgeClassDef(
    BridgeClassType($type),
    constructors: {},
    methods: {
      'position': BridgeMethodDef(
        BridgeFunctionDef(
          returns: BridgeTypeAnnotation(
            BridgeTypeRef.spec(
              BridgeTypeSpec('dart:core', 'Future'),
            ),
            nullable: true,
          ),
        ),
        isStatic: true,
      ),
    },
    getters: {},
    setters: {},
    fields: {},
    wrap: true,
  );

  @override
  $Value? $getProperty(Runtime runtime, String identifier) {
 
    switch (identifier) {
      case 'position':
        return $Future.wrap($value, (val) => $Point.wrap(val));
      default:
        return _superclass.$getProperty(runtime, identifier);
    }
  }

  @override
  get $reified => throw UnimplementedError();

  @override
  int get $runtimeType => throw UnimplementedError();

  @override
  void $setProperty(Runtime runtime, String identifier, $Value value) =>
      throw UnimplementedError();
}


```

I'm also setting up a global class so I can easily make calls to the compiler when the user submits their code. 
Kinda lost here too. 

```dart
import 'package:dart_eval/dart_eval.dart';

import '../../bridges/flutter_auto_gui_bridge.dart';

class FlutterAutoGUIService {
  final compiler = Compiler();
  FlutterAutoGUIService() {
    compiler.defineBridgeClass(
      $FlutterAutoGUI.$declaration,
    );

    Future<void> execute(String code) async {
      final program = compiler.compile(
        {
          'automator': {
            'automator.dart': code,
          },
        },
      );

      final runtime = Runtime.ofProgram(program)
        ..registerBridgeFunc(library, name, (runtime, target, args) => null);
    }
  }
}

```

**Comment by ethanblake4** [comment, marked as answer]
> Would I have to do this if I have the following:

Nope, you're good.

> Would that be something like this?

No, those are instance runtime bindings. You can see examples of static runtime bindings [here](https://github.com/ethanblake4/dart_eval/blob/39f4c0df8fb1c506f45a0445db9f13ddef1aab3c/example/dart_eval_example.dart#L73) (which references [this](https://github.com/ethanblake4/dart_eval/blob/39f4c0df8fb1c506f45a0445db9f13ddef1aab3c/example/dart_eval_example.dart#L130)) and [here](https://github.com/ethanblake4/flutter_eval/blob/ccb607ef6932dc60139e94fe8e27cd48c4589135/lib/flutter_eval.dart#L198) (which references [this](https://github.com/ethanblake4/flutter_eval/blob/ccb607ef6932dc60139e94fe8e27cd48c4589135/lib/src/material/scaffold.dart#L87)). Constructors and static methods are defined basically the same way at runtime. (**Edit:** I did notice you already started doing this in your global class)

> Do I need to define these overloads for the static functions?

No.

> Finally this is what my file now looks like

I think you can remove `nullable: true` with the Future type reference.

> I'm also setting up a global class so I can easily make calls to the compiler when the user submits their code.
Kinda lost here too.

I'd recommend making `setupFlutterAutoGuiForCompile()` and `setupFlutterAutoGuiForRuntime()` functions like [flutter_eval has](https://github.com/ethanblake4/flutter_eval/blob/master/lib/flutter_eval.dart). Then, you can sort of just replicate the first example in the [Advanced usage](https://github.com/ethanblake4/flutter_eval#advanced-usage) section of flutter_eval, though it doesn't have to be in a widget. Note that you will have to add flutter_eval itself as a dependency (in your normal pubspec) and call `setupFlutterForCompile()`/`setupFlutterForRuntime()` (in addition to your own setup functions) to make the compiler/runtime aware of the Flutter SDK classes.

**Reply by Chappie74** [reply]
Thank you for the flutter_eval example. Really helped to make sense of what's going on. 
I'll work with these tips for now and see how far I get.

**Reply by Chappie74** [reply]
Hey so one thing. I'd essentially have to wait until you release a version with the binding for Point etc?

![image](https://user-images.githubusercontent.com/21108487/190834560-24b9489d-79cc-4674-93ea-82456905e46c.png)

**Reply by ethanblake4** [reply]
Yeah unless you write those bindings yourself. I'll finish it in the next day or two though, already have some of it done

**Reply by Chappie74** [reply]
much appreciated, thanks.

**Reply by ethanblake4** [reply]
Just released v0.4.7 versions of dart_eval and flutter_eval which should have the bindings you need :)

**Reply by Chappie74** [reply]
Wonderful!! Will take a look.

**Reply by Chappie74** [reply]
Looks good. Taking a look at the implementation of $Point has helped me get an even better understand on how to write binding. Thanks again.

One question, when it comes to enums I've defined, would it basically be  exactly the same thing?
Do you have any examples I can look at?

**Reply by ethanblake4** [reply]
Yeah flutter_eval has some enums. Check out FontWeight -  [definition](https://github.com/ethanblake4/flutter_eval/blob/a2072502b1c8ce583c36d3aa696036b615070253/lib/src/sky_engine/ui/text.dart#L35), [compiler](https://github.com/ethanblake4/flutter_eval/blob/a2072502b1c8ce583c36d3aa696036b615070253/lib/flutter_eval.dart#L113), [runtime](https://github.com/ethanblake4/flutter_eval/blob/a2072502b1c8ce583c36d3aa696036b615070253/lib/flutter_eval.dart#L246)

**Reply by Chappie74** [reply]
Hey shouldn't the return type  [here](https://github.com/ethanblake4/dart_eval/blob/3cb310cba16f9dea0bbbabeed4a85d5eaff48fb7/lib/src/eval/shared/stdlib/math/point.dart#L132) be $double?

**Reply by ethanblake4** [reply]
It doesn't matter, it would get wrapped in a $Function anyway and lose static type information. dart_eval reimplements the Dart type system itself, it doesn't (and can't) really make use of the builtin one.

**Reply by Chappie74** [reply]
`.` after Point [here](https://github.com/ethanblake4/dart_eval/blob/3cb310cba16f9dea0bbbabeed4a85d5eaff48fb7/lib/src/eval/shared/stdlib/math/point.dart#L17)  is a typo?

**Reply by ethanblake4** [reply]
Nope, that [indicates the default constructor](https://pub.dev/packages/flutter_eval#calling-functions-and-passing-arguments) when bridging

**Reply by Chappie74** [reply]
Ahh gotcha. Thanks.

**Comment by Chappie74** [comment]
@ethanblake4 Sorry but I'm back again haha. 

So. This is what I've managed so far. I think everything here should be fine (focusing more on `$position` function). 

```dart
import 'package:dart_eval/dart_eval.dart';
import 'package:dart_eval/dart_eval_bridge.dart';
import 'package:dart_eval/stdlib/core.dart';
import 'package:dart_eval/stdlib/math.dart';
import 'package:flutter_auto_gui/flutter_auto_gui.dart';

const _flutterAutoGUISource = '''
export 'package:flutter_auto_gui/flutter_auto_gui.dart' show FlutterAutoGUI;
''';

class $FlutterAutoGUI extends $Instance {
  static const $type = BridgeTypeRef.spec(
    BridgeTypeSpec(
      'package:flutter_auto_gui/flutter_auto_gui.dart',
      'FlutterAutoGUI',
    ),
  );

  @override
  final FlutterAutoGUI $value;

  /// Configure this class for compilation in a [Compiler].
  static void configureForCompile(Compiler compiler) {
    compiler.defineBridgeClass($declaration);
    compiler.addSource(DartSource(
        'package:flutter_auto_gui/flutter_auto_gui.dart',
        _flutterAutoGUISource));
  }

  /// Configure this class for runtime in a [Runtime].
  static void configureForRuntime(Runtime runtime) {
    runtime.registerBridgeFunc(
      'package:flutter_auto_gui/flutter_auto_gui.dart',
      'FlutterAutoGUI.position',
      $FlutterAutoGUI.$position,
    );
  }

  $FlutterAutoGUI.wrap(this.$value);

  static const $declaration = BridgeClassDef(
    BridgeClassType($type),
    constructors: {},
    methods: {
      'position': BridgeMethodDef(
        BridgeFunctionDef(
          returns: BridgeTypeAnnotation(
            BridgeTypeRef.spec(
              BridgeTypeSpec('dart:core', 'Future'),
            ),
          ),
        ),
        isStatic: true,
      ),
    },
    getters: {},
    setters: {},
    fields: {},
    wrap: true,
  );

  static $Future $position(
      Runtime runtime, $Value? target, List<$Value?> args) {
    return $Future.wrap(
      FlutterAutoGUI.position(),
      (value) => $Point.wrap(value),
    );
  }

  @override
  $Value? $getProperty(Runtime runtime, String identifier) {
    throw UnimplementedError();
  }

  @override
  get $reified => throw UnimplementedError();

  @override
  int get $runtimeType => throw UnimplementedError();

  @override
  void $setProperty(Runtime runtime, String identifier, $Value value) =>
      throw UnimplementedError();
}
```

I've went ahead and created static function for the compile and runtime set ups like I've seen you done in flutter_eval (i like this approach)
I'm a bit confused with the compile.addSource bit though. 
So the URI in the DartSource constructor is the uri to the package, I get that. And the other field is basically the source code?
is what i'm doing here correct? take a look at `_flutterAutoGUISource`

**Reply by ethanblake4** [reply]
Looks good. You don't need to do `addSource` unless you want to accurately represent your internal package URIs for the sake of code cleanliness. If you do want to, you would put:

`export 'package:flutter_auto_gui/src/flutter_auto_gui.dart' ...'`

And then you would change the `$type` spec to also have `src` in it's URI. (edit: and also the registerBridgeFunc())

**Reply by ethanblake4** [reply]
One more thing is that `$reified` should always return `$value`. Can't say if it would break without it but either way it's good practice.
